### PR TITLE
Support Window Node

### PR DESCRIPTION
**Pull request by junjiejiangjjj**
Signed-off-by: junjie.jiang <junjie.jiang@zilliz.com>

### CODE REVIEW

**Review by GuoRentong — CHANGES_REQUESTED** [review]

File: towhee/runtime/data_queue.py [hunk 1]
```diff
@@ -199,6 +220,9 @@ def get_col_type(self, index):
         return self._cols[index].col_type
 
 
+class _Empty:
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply towhee/runtime/data_queue.py:252]: `None` may be the legal data in the queue.

File: towhee/runtime/nodes/_window.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        ret = {}
+        for i in range(len(self._schema)):
+            ret[self._schema[i]] = cols[i]
+        return ret
+
+    def process_step(self) -> bool:
+        """
+        Process each window data.
+        """
+        datas = self._get_buffer()
```
[the remark this answers is not in the document]
Comment (junjiejiangjjj) [reply towhee/runtime/nodes/_window.py]: done

File: towhee/runtime/data_queue.py [hunk 3]
```diff
@@ -52,30 +57,55 @@ def put(self, inputs: Union[Tuple, List]) -> bool:
 
             for i in range(len(inputs)):
                 self._data[i].put(inputs[i])
-            self._size += 1
-            self._not_empty.notify()
+
+            new_size = self._get_size()
+            inc_size = new_size - self._size
+            self._size = new_size
+            if inc_size > 0:
```
Comment (GuoRentong) [inline towhee/runtime/data_queue.py:64]: if self._size > 0?

File: towhee/runtime/data_queue.py [hunk 4]
```diff
@@ -113,14 +143,22 @@ def col_size(self) -> int:
 
     def clear_and_seal(self):
```
Comment (GuoRentong) [inline towhee/runtime/data_queue.py:144]: in which case do we need this method?

File: towhee/runtime/data_queue.py [hunk 5]
```diff
@@ -156,7 +194,17 @@ def type_schema(self) -> List[str]:
             >>> dq.type_schema
             [<ColumnType.SCALAR: 2>, <ColumnType.QUEUE: 1>]
         """
-        return list(self._schema.col_types())
+        return self._schema.col_types
+
+    def _get_size(self):
+        for index in self._scalar_index:
```
Comment (GuoRentong) [inline towhee/runtime/data_queue.py:200]: we don't need to check this condition every time after all the scalar value being set

**Review by GuoRentong — CHANGES_REQUESTED** [review]

File: towhee/runtime/data_queue.py (hunk 4 above)
Comment (junjiejiangjjj) [reply towhee/runtime/data_queue.py:144]: when node failed, clear all data and stop running

Comment (junjiejiangjjj) [reply towhee/runtime/data_queue.py:144]: When failed, clear all datas and exit

File: towhee/runtime/data_queue.py (hunk 3 above)
Comment (junjiejiangjjj) [reply towhee/runtime/data_queue.py]: done

**Pull request merged**